perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_GIDRanges.java

```diff
@@ -0,0 +1,96 @@
+import cz.metacentrum.perun.core.api.AttributeDefinition;
+import cz.metacentrum.perun.core.api.AttributesManager;
+import cz.metacentrum.perun.core.api.Facility;
+import cz.metacentrum.perun.core.api.exceptions.AttributeNotExistsException;
+import cz.metacentrum.perun.core.api.exceptions.ConsistencyErrorException;
+import cz.metacentrum.perun.core.api.exceptions.InternalErrorException;
+import cz.metacentrum.perun.core.api.exceptions.WrongAttributeAssignmentException;
+import cz.metacentrum.perun.core.api.exceptions.WrongAttributeValueException;
+import cz.metacentrum.perun.core.api.exceptions.WrongReferenceAttributeValueException;
+import cz.metacentrum.perun.core.impl.PerunSessionImpl;
+import cz.metacentrum.perun.core.impl.Utils;
+import cz.metacentrum.perun.core.implApi.modules.attributes.FacilityVirtualAttributesModuleAbstract;
+import cz.metacentrum.perun.core.implApi.modules.attributes.FacilityVirtualAttributesModuleImplApi;
+
+/**
+ * GID Ranges computed for specific namespace on chosen facility
+ *
+ * @author Michal Stava <stavamichal@gmail.com>
+ */
+public class urn_perun_facility_attribute_def_virt_GIDRanges extends FacilityVirtualAttributesModuleAbstract implements FacilityVirtualAttributesModuleImplApi {
+
+	public void checkAttributeValue(PerunSessionImpl sess, Facility facility, Attribute attribute) throws InternalErrorException, WrongAttributeValueException, WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
+		try {
+			Attribute gidNamespaceAttribute = getUnixGIDNamespaceAttribute(sess, facility);
+			if(gidNamespaceAttribute.getValue() == null) throw new WrongReferenceAttributeValueException(attribute, gidNamespaceAttribute, "There is missing GID namespace on the facility.");
+			Attribute namespaceGIDRangesAttribute = getNamespaceGIDRangesAttribute(sess, (String) gidNamespaceAttribute.getValue());
+			sess.getPerunBl().getAttributesManagerBl().checkAttributeValue(sess, (String) gidNamespaceAttribute.getValue(), namespaceGIDRangesAttribute);
+		} catch(WrongReferenceAttributeValueException ex) {
+			throw new WrongReferenceAttributeValueException(attribute, ex.getReferenceAttribute());
+		}
+	}
+
+	public Attribute getAttributeValue(PerunSessionImpl sess, Facility facility, AttributeDefinition attributeDefinition) throws InternalErrorException {
+		Attribute attribute = new Attribute(attributeDefinition);
+		Attribute gidNamespaceAttribute = getUnixGIDNamespaceAttribute(sess, facility);
+		if(gidNamespaceAttribute.getValue() == null) return attribute;
+		Attribute namespaceGIDRangesAttribute = getNamespaceGIDRangesAttribute(sess, (String) gidNamespaceAttribute.getValue());
+		attribute = Utils.copyAttributeToVirtualAttributeWithValue(namespaceGIDRangesAttribute, attribute);
+		return attribute;
+	}
+
+	public boolean setAttributeValue(PerunSessionImpl sess, Facility facility, Attribute attribute) throws InternalErrorException, WrongReferenceAttributeValueException {
```

Comment:
I would rather not implement this method. This can be called only by perun admin and I would recommend to require to set entityless attribute directly (to avoid possible mistake, when one could thing he only changes settings for this facility).

Reply:
We discussed this with @stavamichal and this implementation differ from module default also in original GID module. I believe it throws exception, if you are trying to set different value, than is already stored and return false otherwise. In both cases it doesn't store/change anything. We are not sure, if exception is not expected somewhere in peruns logic. Like some automatic process wants to set required attributes, it sets value inside the object and call setAttribute() method. If it would only return false, process could continue with wrong value in object causing problems to other attributes. Like this it kill the process. But this code might not be relevant anymore. What do you think @licehammer ?

Reply:
You are right Pavel. I just saw the implementation of the method itself and I didn't check what is does exactly. My bad. This method can stay as it is.

To your question: This method doesn't need to be implemented, because when you don't have the "fill" method, no automatic process should try to set this attribute. On the other hand, user might do that for example by modifying some attributes from GUI and then select and save all (including this one). In that case might be handy to have this method implemented as it is now.